mirascope/core/base/prompt.py

```diff
@@ -345,7 +351,138 @@ def _is_function(
     return isinstance(prompt, types.FunctionType)
 
 
-def prompt_template(template: str) -> PromptDecorator:
+WrappedMessagesAsyncFunction: TypeAlias = Callable[_P, Awaitable[Messages]]
+WrappedMessagesSyncFunction: TypeAlias = Callable[_P, Messages]
+
+MessagesSyncFunction: TypeAlias = Callable[_P, _MessageFuncReturnT]
+MessagesAsyncFunction: TypeAlias = Callable[_P, Awaitable[_MessageFuncReturnT]]
+
+
+class MessagesDecorator(Protocol):
+    @overload
+    def __call__(
+        self,
+        messages_fn: MessagesSyncFunction[_P, _MessageFuncReturnT],
+    ) -> WrappedMessagesSyncFunction[_P]: ...
+
+    @overload
+    def __call__(
+        self,
+        messages_fn: MessagesAsyncFunction[_P, _MessageFuncReturnT],
+    ) -> WrappedMessagesAsyncFunction[_P]: ...
+
+    def __call__(
+        self,
+        messages_fn: MessagesSyncFunction[_P, _MessageFuncReturnT]
+        | MessagesAsyncFunction[_P, _MessageFuncReturnT],
+    ) -> WrappedMessagesSyncFunction[_P] | WrappedMessagesAsyncFunction[_P]: ...
+
+
+def _is_message_param_function(
+    messages_fn: MessagesSyncFunction[_P, _MessageFuncReturnT]
+    | MessagesAsyncFunction[_P, _MessageFuncReturnT]
+    | None
+    | str,
+) -> TypeIs[
+    MessagesSyncFunction[_P, _MessageFuncReturnT]
+    | MessagesAsyncFunction[_P, _MessageFuncReturnT]
+]:
+    return isinstance(messages_fn, types.FunctionType)
+
+
+def _parse_messages(value: object, attrs: dict[str, Any]) -> Messages:
+    if isinstance(value, str):
+        if parsed_message := parse_content_template("user", value, attrs):
+            return [parsed_message]
+        return []
+    if isinstance(value, BaseMessageParam):
+        return [value]
+    if isinstance(value, list):
+        if all(isinstance(item, BaseMessageParam) for item in value):
+            return value
+        if all(isinstance(item, str) for item in value):
+            if parsed_messages := parse_content_template("user", value, attrs):
+                return [parsed_messages]
+            return []
+        # How to detect non-text types?
+
+    raise ValueError(f"Invalid message type: {type(value)} for value: {value}")
```

Comment:
My understanding is that it is difficult to determine the type of content (`previous_book `) from the examples you raised.
How can I make a decision other than `str`?
```python
def recommend_book(previous_book: Image.Image) -> Messages:
    return ["I just read this book:", previous_book, "What should I read next?"]
    # return BaseMessageParam(role='user', content=["I just read this book:", previous_book, "What should I read next?"])
```